### PR TITLE
Refactor local-only `Condition::Resharding` into serializable `Condition::HashRing`

### PR DESCRIPTION
**Pull request by ffuugoo**
This PR refactor local-only `hash_ring::Filter` and `Condition::Resharding` into a generic, serializable `Condition::HashRing`. This _(or some specialized API)_ is required to move resharding driver into external service.

`Condition::HashRing` seems like a generally useful API to test/debug resharding correctness, so it was decided to implement it, instead of a simpler/specialized solution.

__TODO:__
- [x] ~~documentation~~
  - we're keeping it hush-hush 🤫
  - [ ] ~~Document fields of `HashRingCondition`~~
  - [ ] ~~OpenAPI~~
  - [ ] ~~gRPC~~
- [x] unit-tests
- [ ] ~~intergraiton tests~~
  - integration tests would be added in a follow up PR, *once we expose `Commit*HashRing` API*, because it would be trivial to *setup* `HashRingCondition` tests once `Commit*HashRing` API is available

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by ffuugoo** [comment]
Sorry for a single >500 LoC commit, but this was **so damn annoying** to implement, I didn't have enough energy to do a proper commit history. 😬

Opened as draft, cause I still have to finish some minor things here and there and add some tests (and PR description), but the bulk of PR is ready, I think.

**Comment by ffuugoo** [comment]
@generall I think the only thing left for this PR is to add some documentation to `HashRingCondition` fields, I'll do so once I'm back from a break. Other than that, it's ready for review.

File: lib/segment/src/types.rs [hunk 1]
```diff
@@ -1972,7 +1970,102 @@ impl NestedCondition {
...
+    /// Hash ring to hash point IDs with
+    pub ring: scaled_hashring::ScaledHashRing<ShardId>,
+
+    /// Shard IDs that hashed point IDs should match
+    pub match_shard_ids: HashSet<ShardId>,
+
+    /// Marks if this condition should *not* be transferred to remote shards when applying queries
+    // or operations to replica set
+    #[serde(skip)]
+    pub is_local_only: bool,
```
Comment (generall) [inline lib/segment/src/types.rs:1985]: Segment should not know anything about replica sets

**Comment by ffuugoo** [comment]
Closing, cause we ended up with a different approach in #4928.

**Pull request closed without merging**